finesse/gui/hardware_set/hardware_set.py

```diff
@@ -0,0 +1,73 @@
+"""The HardwareSet dataclass and associated helper functions."""
+from __future__ import annotations
+
+import logging
+from collections.abc import Generator
+from dataclasses import dataclass, field
+from importlib import resources
+from pathlib import Path
+from typing import Any
+
+import yaml
+from frozendict import frozendict
+
+from finesse.device_info import DeviceInstanceRef
+from finesse.gui.device_connection import close_device, open_device
+
+
+@dataclass(frozen=True)
+class OpenDeviceArgs:
+    """Arguments needed to open a device."""
+
+    instance: DeviceInstanceRef
+    class_name: str
+    params: frozendict[str, Any] = field(default_factory=frozendict)
+
+    def open(self) -> None:
+        """Open the device."""
+        open_device(self.class_name, self.instance, self.params)
+
+    def close(self) -> None:
+        """Close the device."""
+        close_device(self.instance)
+
+    @staticmethod
+    def create(
+        instance: str, class_name: str, params: dict[str, Any] = {}
+    ) -> OpenDeviceArgs:
+        """Create an OpenDeviceArgs using basic types."""
+        return OpenDeviceArgs(
+            DeviceInstanceRef.from_str(instance), class_name, frozendict(params)
+        )
+
+
+@dataclass(frozen=True)
+class HardwareSet:
+    """Represents a collection of devices for a particular hardware configuration."""
+
+    name: str
+    devices: frozenset[OpenDeviceArgs]
+    file_path: Path
+    read_only: bool
+
+    @staticmethod
+    def load(file_path: Path, read_only: bool = False) -> HardwareSet:
```

Comment:
As above, I would personally use a `classmethod` and not an `staticmethod` for this, but up to you.